File: exercises/leetcode/algorithms/1167-minimum-cost-to-connect-sticks/solution.py
class Solution:
    def connectSticks(self, sticks: List[int]) -> int:
        # Convert the sticks list into a min-heap
        heapq.heapify(sticks)
        cost = 0

        # Continue until there's only one stick left
        while len(sticks) > 1:
            first = heapq.heappop(sticks)   # Get the smallest stick
            second = heapq.heappop(sticks)  # Get the second smallest stick

            # Combine the two sticks
            combined = first + second
            cost += combined

            # Add the combined stick back into the min-heap
            heapq.heappush(sticks, combined)

        return cost

File: exercises/leetcode/algorithms/1167-minimum-cost-to-connect-sticks/solution.ts
function connectSticks(sticks: number[]): number {
   const minHeap = sticks.sort((a, b) => a - b)
   let cost = 0

   // Continue until there's only one stick left
   while (minHeap.length > 1) {
      const first = minHeap.shift() // Get the smallest stick
      const second = minHeap.shift() // Get the second smallest stick

      // Combine the two sticks
      const combined = first + second
      cost += combined

      addToSortedArray(minHeap, combined)
   }

   return cost
}

function addToSortedArray(array: number[], item: number) {
   let i = 0
   while (i < array.length && array[i] < item) {
      i++
   }
   array.splice(i, 0, item)
}
